File: app-ios/ios/controller/ArticleTableViewController.m
//
//  ArticleTableViewController.m
//  ios
//
//  Created by Peter Zhan on 2017/10/15.
//  Copyright © 2017年 com.enuocms. All rights reserved.
//

#import "ArticleTableViewController.h"
#import "ArticleService.h"

@interface ArticleTableViewController ()
@property (strong,nonatomic) ArticleService* articleService;
@property (strong,nonatomic) APIPageResult* pageList;
@end

@implementation ArticleTableViewController
- (ArticleService*) articleService{
    if(_articleService==nil){
        _articleService = [[ArticleService alloc] init];
    }
    return _articleService;
}

- (APIPageResult*) pageList{
    if(_pageList==nil){
        _pageList = [[APIPageResult alloc] init];
        _pageList.pageSize = 10;
        _pageList.pageIndex = 1;
    }
    return _pageList;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    __weak typeof(self) vc = self;
    //下拉刷新
    self.tableView.header = [MJRefreshNormalHeader headerWithRefreshingBlock:^{
        [vc loadData:[NSNumber numberWithBool:NO]];
    }];
    
    //加载数据
    [self loadData:[NSNumber numberWithBool:NO]];
    
    //上拉刷新
    self.tableView.footer = [MJRefreshBackNormalFooter footerWithRefreshingBlock:^{
        [vc loadData:[NSNumber numberWithBool:YES]];
    }];
    
}

-(void) loadData:(NSNumber*) more{
    if([more boolValue]){
        self.pageList.pageIndex =  self.pageList.pageIndex + 1;
    }else{
        self.pageList.pageIndex =  1;
        [self.pageList.items removeAllObjects];
    }
    
    __weak typeof(self) vc = self;
    MBProgressHUD *hud = [Common showHUDAddedTo:self.view animated:YES];
    [self.articleService pageList:self.pageList.pageIndex pageSize:self.pageList.pageSize success:^(APIResult *result, APIPageResult *obj) {
        if(result.success){
            [vc.pageList.items addObjectsFromArray:obj.items];
            vc.pageList.total = obj.total;
        }else{
            [vc.view makeToast:result.msg duration:0.3 position:CSToastPositionTop ];
        }
        [hud hideAnimated:NO];
        //header
        if(![more boolValue]){
              [vc.tableView.header endRefreshing];
        }else{
            //footer
            if([vc.pageList.items count] >= vc.pageList.total){
                [vc.tableView.footer endRefreshing];
                [vc.tableView.footer setHidden:YES];
            }else{
                [vc.tableView.footer endRefreshing];
                [vc.tableView.footer setHidden:NO];
            }
        }
        [vc.tableView reloadData];
    } failure:^(NSError *error) {
        [hud hideAnimated:NO];
    }];
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
     return [self.pageList.items count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"article_cell" forIndexPath:indexPath];
    Categorys *ent = [self.pageList.items objectAtIndex:indexPath.row];
    UILabel *textLabel = [cell.contentView viewWithTag:11];
    [textLabel setText:ent.title];
    //[cell.textLabel setText:ent.title];
    return cell;
}


/*
// Override to support conditional editing of the table view.
- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the specified item to be editable.
    return YES;
}
*/

/*
// Override to support editing the table view.
- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        // Delete the row from the data source
        [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
    } else if (editingStyle == UITableViewCellEditingStyleInsert) {
        // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
    }   
}
*/

/*
// Override to support rearranging the table view.
- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath {
}
*/

/*
// Override to support conditional rearranging of the table view.
- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the item to be re-orderable.
    return YES;
}
*/

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: app-swift/ios/ios/controller/ArticleTableViewController.swift
//
//  ArticleTableViewController.swift
//  ios
//
//  Created by Peter Zhan on 2017/10/18.
//  Copyright © 2017年 com.enuocms. All rights reserved.
//

import UIKit
import MJRefresh

class ArticleTableViewController: UITableViewController {
    
    let articleService = ArticleService.init()
    var pageList  =  ApiPageResult.init(1,10)
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.tableView.header = MJRefreshNormalHeader.init(refreshingBlock: {
            self.loadData(false)
        })
        
        self.tableView.footer = MJRefreshBackNormalFooter.init(refreshingBlock: {
             self.loadData(true)
        })
        
        self.loadData(false)
    }
    
    func loadData(_ more:Bool) {
        
        if(more){
            pageList.pageIndex = pageList.pageIndex + 1
        }else{
             pageList.pageIndex = 1
            pageList.items.removeAll()
        }
        
        let hud = Common.showHUDAddedTo(view: self.view, animated: true)
        articleService.pagelist(pageIndex: pageList.pageIndex, pageSize: pageList.pageSize, success: { (result, obj) in
            if(result.success){
                self.pageList.total = obj.total
                for item in obj.items{
                    self.pageList.items.append(item)
                }
                self.tableView.reloadData()
            }
            
            if(more){
                self.tableView.footer.endRefreshing()
                if(self.pageList.items.count >= self.pageList.total){
                    self.tableView.footer.isHidden = true
                }else{
                    self.tableView.footer.isHidden = false
                }
            }else{
                self.tableView.header.endRefreshing()
            }
            
            hud.hide(animated: false)
        }, failure: { (error) in
              hud.hide(animated: false)
        })
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return pageList.items.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "article_cell", for: indexPath)
        let ent = pageList.items[indexPath.row] as! Article
        let label = cell.contentView.viewWithTag(11) as! UILabel
        label.text = ent.title
        //cell.textLabel?.text = ent.title
        return cell
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
}
